### PR TITLE
CMake: Make it possible to override the scale factor

### PR DESCRIPTION
**Pull request by tronical**

### CODE REVIEW

**Comment by tronical** [comment]
This is an alternative to #5016 . If that looks alright, I can document this and also add the touch scale factor fixes from #5016 on top.

**Review by ogoffart — COMMENTED** [review]
I think we need to improve a bit. As we have the following stuff to control the scale factor:

 - At cmake configure time **of Slint itself**: the `SLINT_SCALE_FACTOR_OVERRIDE` added in this PR, which then control the default scale factor an application has if the actual platform/backend don't specify it.
 - At compile time of the application, the slint_compiler uses the `SLINT_SCALE_FACTOR` env variable when generating images and font.
 - At application runtime, our winit backend uses the `SLINT_SCALE_FACTOR` env variable to override the actual scale factor from the display.


I think it really is not ideal that there could be two different choice when configuring slint itself and when building the app. I wonder how we can improve on that.

One idea would be to add the `dispatch_event(ScaleFactorChanged)` in the generated code. Not sure if that works.

File: internal/compiler/generator/rust.rs [hunk 1]
```diff
@@ -326,6 +330,15 @@ fn generate_shared_globals(llr: &llr::CompilationUnit) -> TokenStream {
         .map(global_inner_name)
         .collect::<Vec<_>>();
 
+    let apply_constant_scale_factor = if !compiler_config.const_scale_factor.approx_eq(&1.0) {
+        let factor = compiler_config.const_scale_factor;
+        Some(
+            quote!(adapter.window().dispatch_event(slint::platform::WindowEvent::ScaleFactorChanged{ scale_factor: #factor as f32 });),
+        )
```
Comment (ogoffart) [inline internal/compiler/generator/rust.rs:337]: Not important, but i'd suggest to move the `as f32` in slint compilation time from rust compilation time.

```suggestion
        let factor = compiler_config.const_scale_factor as f32;
        Some(
            quote!(adapter.window().dispatch_event(slint::platform::WindowEvent::ScaleFactorChanged{ scale_factor: #factor });),
        )
```

also this `else{None}` pattern is maybe something witch could be done with `(!compiler_config.const_scale_factor.approx_eq(&1.0)).then(|| ...)`  you decide what takes the less lines :-)

File: tools/compiler/main.rs [hunk 2]
```diff
@@ -47,6 +47,10 @@ struct Cli {
     #[arg(long, name = "style name", action)]
     style: Option<String>,
 
+    /// The constant scale factor to apply
```
Comment (ogoffart) [inline tools/compiler/main.rs:50]:
```suggestion
    /// The scale factor for embedded assets and set by default on the window
```

File: api/cpp/cmake/SlintMacro.cmake [hunk 3]
```diff
@@ -9,6 +9,7 @@ set_property(CACHE DEFAULT_SLINT_EMBED_RESOURCES PROPERTY STRINGS
 ## This requires CMake 3.23 and does not work in 3.26 AFAICT.
 # define_property(TARGET PROPERTY SLINT_EMBED_RESOURCES
 #     INITIALIZE_FROM_VARIABLE DEFAULT_SLINT_EMBED_RESOURCES)
+set(DEFAULT_SLINT_SCALE_FACTOR "" CACHE STRING "A constant scale factor to apply. Use only in MCU environments")
```
Comment (ogoffart) [inline api/cpp/cmake/SlintMacro.cmake:12]: why is this `DEFAULT_`?  shouldn't all these var start by `SLINT_` so it should be `SLINT_DEFAULT_SCALE_FACTOR` or just `SLINT_SCALE_FACTOR` or perhaps `SLINT_CONST_SCALE_FACTOR`

File: tests/screenshots/build.rs [hunk 4]
```diff
@@ -107,18 +107,13 @@ fn main() -> std::io::Result<()> {
         generate_source(source.as_str(), &mut output, testcase, scale_factor.unwrap_or(1.))
             .unwrap();
 
-        let scale_factor = scale_factor.map_or(String::new(), |scale_factor| {
-            format!("slint::platform::WindowAdapter::window(&*window).dispatch_event(slint::platform::WindowEvent::ScaleFactorChanged {{ scale_factor: {scale_factor}f32 }});")
-        });
```
Comment (ogoffart) [inline tests/screenshots/build.rs:112]: I like the fact that this suddenly makes this feature tested :-)

**Review by ogoffart — COMMENTED** [review]
(i guess the missing docuentation is one reason this is still a draft)

File: internal/compiler/generator/rust.rs (hunk 1 above)
Comment (tronical) [reply internal/compiler/generator/rust.rs:337]: I was contemplating that pattern, but I don't find `(!cond)..then(|| ...)` quite readable :-(. Good point about the f32.

File: api/cpp/cmake/SlintMacro.cmake (hunk 3 above)
Comment (tronical) [reply api/cpp/cmake/SlintMacro.cmake:12]: It's the same as with the resource embedding: This cache property comes from the application, not Slint. We can discuss with @hunger more, but if you're unsure I can also remove this fallback for now.

**Comment by tronical** [comment]
> (i guess the missing docuentation is one reason this is still a draft)

Yes. Well, *was* ;-)

File: api/rs/build/lib.rs [hunk 5]
```diff
@@ -162,6 +162,17 @@ impl CompilerConfiguration {
         };
         Self { config }
     }
+
+    /// Sets the scale factor to be applied to all `px` to `phx` conversions
+    /// as constant value. This is only intended for MCU environments. Use
+    /// in combination with [`Self::embed_resources`] to pre-scale images and glyphs
+    /// accordingly.
+    #[must_use]
+    pub fn with_scale_factor(self, factor: f32) -> Self {
```
Comment (ogoffart) [inline api/rs/build/lib.rs:171]: ~~should it perhaps be called `with_const_scale_factor`?~~
(nah, that would send the wrong signal that the application scale factor will be const)

File: api/cpp/docs/cmake_reference.md [hunk 6]
```diff
@@ -48,3 +48,17 @@ This target property is initialised from the global `DEFAULT_SLINT_EMBED_RESOURC
 # Example: when building my_application, specify that the compiler should embed the resources in the binary
 set_property(TARGET my_application PROPERTY SLINT_EMBED_RESOURCES embed-files)
 ```
+
+## Scale Factor for Microcontrollers
+
+When targeting a Microcontroller, there exists no windowing system that provides a device pixel ratio to
+map logical lengths in Slint (`px`) to physical pixels (`phx`). If desired, you can provide this ratio at
+compile time by setting the `SLINT_SCALE_FACTOR` target property on your CMake target.
```
Comment (ogoffart) [inline api/cpp/docs/cmake_reference.md:56]: But there is a platform that is supposed to provide that.

Anyway, i don't know how to improve it.

**Review by ogoffart — APPROVED** [review]

**Review by ogoffart — APPROVED** [review]

**Pull request merged**